### PR TITLE
SAK-50395 Gradebook Migrate from Handsontable to Tabulator

### PR DESCRIPTION
**Pull request by kunaljaykam**
https://sakaiproject.atlassian.net/browse/SAK-50395

Tabulator: https://tabulator.info/

### CODE REVIEW

File: gradebookng/tool/src/webapp/scripts/gradebook-gbgrade-table.js [hunk 1]
```diff
@@ -692,39 +779,32 @@ GbGradeTable.renderTable = function (elementId, tableData) {
...
+      columnType: "studentname",
+      col: "0"
+    },
     width: 220,
-    sortCompare: function(a, b) {
-        return GbGradeTable.studentSorter(a, b);
-    }
+    // sorter: function(a, b) {
+    //     return GbGradeTable.studentSorter(a, b);
+    // }
```
Comment (ern) [inline gradebookng/tool/src/webapp/scripts/gradebook-gbgrade-table.js:793]:
```suggestion
```

File: gradebookng/tool/src/webapp/scripts/gradebook-gbgrade-table.js [hunk 2]
```diff
@@ -799,536 +906,524 @@ GbGradeTable.renderTable = function (elementId, tableData) {
...
+  }
+ 
+  });
 
-    // This function is another hotspot.  Efficiency is paramount!
-    afterGetColHeader: function(col, th) {
-      var $th = $(th);
+//   GbGradeTable.instance.on("cellEditing", function(cell) {
+//     const checkAndDisableEditor = function() {
+//         const cellElement = cell.getElement();
```
Comment (ern) [inline gradebookng/tool/src/webapp/scripts/gradebook-gbgrade-table.js:931]:
```suggestion
```

File: library/src/skins/default/src/sass/modules/tool/gradebook/_gradebook.scss [hunk 3]
```diff
@@ -232,21 +232,21 @@
...
+  //   min-width: 16px;
+  //   height: 20px;
+  //   color: var(--sakai-text-color-dimmed);
+  // }
+  // #gradeTableWrapper .gb-comment-notification:before,
+  // #gradeTableWrapper .gb-course-comment-notification:before,
+  // .gb-flag-comment:before {
+  //   font-family: "bootstrap-icons";
+  //   content: '\F24B';
+  // }
```
Comment (ern) [inline library/src/skins/default/src/sass/modules/tool/gradebook/_gradebook.scss:249]:
```suggestion
```

**Comment by ern** [comment]
Remove commented out code and remove unnecessary console.log

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/actions/MoveAssignmentRightAction.java [hunk 4]
```diff
@@ -36,6 +36,7 @@ public ActionResponse handleEvent(final JsonNode params, final AjaxRequestTarget
 
 		final Long assignmentId = Long.valueOf(params.get("assignmentId").asText());
 
+
```
Comment (jesusmmp) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/actions/MoveAssignmentRightAction.java:39]:
```suggestion
```

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/component/GbGradeTable.html [hunk 5]
```diff
@@ -277,11 +277,11 @@
             </div>
         </script>
         <script id="sectionsCellTemplate" type="text/template">
-            <div class="gb-value gb-section-cell">
+            <div class="gb-value   d-flex mt-3 justify-content-center gb-section-cell">
```
Comment (jesusmmp) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/component/GbGradeTable.html:280]:
```suggestion
            <div class="gb-value d-flex mt-3 justify-content-center gb-section-cell">
```

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/component/GbGradeTable.html [hunk 6]
```diff
@@ -303,12 +303,14 @@
         </script>
         <script id="cellTemplate" type="text/template">
             <div class="relative">
-                <span class="gb-value">${value}</span>
+                <div class="gb-editable" style="min-height: 20px">
+                    <span class="gb-value   d-flex mt-3 justify-content-center">${value}</span>
```
Comment (jesusmmp) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/component/GbGradeTable.html:307]:
```suggestion
                    <span class="gb-value d-flex mt-3 justify-content-center">${value}</span>
```

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/component/GbGradeTable.html [hunk 7]
```diff
@@ -326,15 +328,15 @@
         </script>
         <script id="courseGradeCellTemplate" type="text/template">
             <div class="relative">
-                <span class="gb-value {if isOverridden} gb-overridden {/if}">
+                <span class="gb-value   d-flex mt-3 justify-content-center {if isOverridden} gb-overridden {/if}">
```
Comment (jesusmmp) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/component/GbGradeTable.html:331]:
```suggestion
                <span class="gb-value d-flex mt-3 justify-content-center {if isOverridden} gb-overridden {/if}">
```

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/model/GbModalWindow.java [hunk 8]
```diff
@@ -148,39 +148,39 @@ private void setDefaultWindowClosedCallback() {
...
-						target.appendJavaScript(String.format("setTimeout(function() {GbGradeTable.selectCourseGradeCell('%s');});",
-								GbModalWindow.this.studentUuidToReturnFocusTo));
-					} else {
-						target.appendJavaScript(String.format("setTimeout(function() {GbGradeTable.selectCell(null, '%s');});",
-								GbModalWindow.this.studentUuidToReturnFocusTo));
-					}
-				} else if (GbModalWindow.this.returnFocusToCourseGrade) {
-					target.appendJavaScript("setTimeout(function() {GbGradeTable.selectCourseGradeCell();});");
-				}
+				// target.appendJavaScript(
```
Comment (jesusmmp) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/model/GbModalWindow.java:151]: Remove commented text

Comment (kunaljaykam) [reply gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/model/GbModalWindow.java:151]: I want to keep it until some testing by QA. After some time I will remove it and probably some codes relateded to this.

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/component/GbGradeTable.html [hunk 9]
```diff
@@ -303,12 +303,15 @@
         </script>
         <script id="cellTemplate" type="text/template">
             <div class="relative">
-                <span class="gb-value">${value}</span>
+                <div class="gb-editable" style="min-height: 20px">
```
Comment (adrianfish) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/component/GbGradeTable.html:306]: This should be in the sass build, not inline.

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/component/GbGradeTable.html [hunk 10]
```diff
@@ -303,12 +303,15 @@
...
+                    <span class="gb-value d-flex mt-3 justify-content-center">${value}</span>
+                </div>
                 <span class="gb-notification"></span>
-                <span class="gb-comment-notification"></span>
-                <div class="btn-group dropend">
+                <span class="gb-comment-notification bi bi-chat-right-dots position-absolute bottom-0" style="right:7px;"></span>
+                <div class="btn-group dropend position-absolute top-0 end-0">
                     <button type="button"
-                        class="btn btn-sm btn-default dropdown-toggle"
+                        class="btn btn-sm btn-default bi bi-gear"
```
Comment (adrianfish) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/component/GbGradeTable.html:313]:
```suggestion
                        class="btn btn-sm btn-default si si-settings"
```

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/component/GbGradeTable.html [hunk 11]
```diff
@@ -326,15 +329,15 @@
         </script>
         <script id="courseGradeCellTemplate" type="text/template">
             <div class="relative">
-                <span class="gb-value {if isOverridden} gb-overridden {/if}">
+                <span class="gb-value d-flex mt-3 justify-content-center {if isOverridden} gb-overridden {/if}">
                     ${value}
                 </span>
                 <span class="gb-notification"></span>
-                <span class="gb-course-comment-notification"></span>
+                <span class="gb-comment-notification bi bi-chat-right-dots position-absolute bottom-0" style="right:7px;"></span>
```
Comment (adrianfish) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/component/GbGradeTable.html:336]: We should not use inline styles.

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/component/GbGradeTable.html [hunk 12]
```diff
@@ -326,15 +329,15 @@
...
                 {if GbGradeTable.settings.isUserAbleToEditAssessments}
-                    <div class="btn-group dropend">
+                    <div class="btn-group dropend position-absolute top-0 end-0">
                         <button type="button"
-                            class="btn btn-sm btn-default dropdown-toggle"
+                            class="btn btn-sm btn-default bi bi-gear"
```
Comment (adrianfish) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/component/GbGradeTable.html:340]:
```suggestion
                            class="btn btn-sm btn-default si si-settings"
```

File: gradebookng/tool/src/webapp/scripts/gradebook-gbgrade-table.js [hunk 13]
```diff
@@ -333,336 +358,391 @@ GbGradeTable.replaceContents = function (elt, newContents) {
...
-
-  if (!valueCell) {
-    valueCell = td.getElementsByClassName('gb-value')[0];
+  const valueCell = td.querySelector('.gb-value');
+  if (valueCell) {
+      if (GbGradeTable.settings.isPercentageGradeEntry && value !== null && value !== "") {
+          GbGradeTable.replaceContents(valueCell, document.createTextNode(`${value}%`));
+      } else {
+          GbGradeTable.replaceContents(valueCell, document.createTextNode(value));
+      }
```
Comment (adrianfish) [inline gradebookng/tool/src/webapp/scripts/gradebook-gbgrade-table.js:402]: We should try and stick to 2 spaces for indentation.

File: gradebookng/tool/src/webapp/scripts/gradebook-gbgrade-table.js [hunk 14]
```diff
@@ -333,336 +358,391 @@ GbGradeTable.replaceContents = function (elt, newContents) {
...
-    $.removeData(td, "categoryId");
+ 
+  onRendered(function() {
 
-    if (GbGradeTable.settings.isPercentageGradeEntry && value != null && value != "") {
-      GbGradeTable.replaceContents(valueCell, document.createTextNode('' + value + '%'));
+    const gradeRubricClass = "gb-grade-rubric-li";
+    const gradeRubricListItem = td.querySelector(`.${gradeRubricClass}`);
+    if (gradeRubricListItem) {
+        gradeRubricListItem.className = (!hasAssociatedRubric || isExternallyMaintained) ? `${gradeRubricClass} d-none` : gradeRubricClass;
```
Comment (adrianfish) [inline gradebookng/tool/src/webapp/scripts/gradebook-gbgrade-table.js:411]:
```suggestion
      gradeRubricListItem.className = (!hasAssociatedRubric || isExternallyMaintained) ? `${gradeRubricClass} d-none` : gradeRubricClass;
```

File: gradebookng/tool/src/webapp/scripts/gradebook-gbgrade-table.js [hunk 15]
```diff
@@ -333,336 +358,391 @@ GbGradeTable.replaceContents = function (elt, newContents) {
...
-          dropdownToggle[0].style.display = 'block';
-          dropdownToggle.attr('aria-hidden', 'false');
-          var dropdownToggleTooltip = GbGradeTable.templates.gradeMenuTooltip.process();
-          dropdownToggleTooltip = dropdownToggleTooltip.replace('{0}', student.firstName + ' ' + student.lastName);
-          dropdownToggleTooltip = dropdownToggleTooltip.replace('{1}', column.title);
-          dropdownToggle.attr('title', dropdownToggleTooltip);
-      }
+    const gradeRubricOption = td.querySelector(".gb-grade-rubric")?.parentElement;
+    if (gradeRubricOption) {
+        gradeRubricOption.classList.toggle("invisible", !hasAssociatedRubric);
```
Comment (adrianfish) [inline gradebookng/tool/src/webapp/scripts/gradebook-gbgrade-table.js:416]:
```suggestion
      gradeRubricOption.classList.toggle("invisible", !hasAssociatedRubric);
```

**Review by adrianfish — CHANGES_REQUESTED** [review]
I've left a few comments. I've not reviewed the entire PR yet but I didn't want to just leave a load of comments all in one bunch. One thing to watch out for is inconsisten indentation. We're using 2 spaces in all the webcomponents and I'm pretty sure the gb js mainly uses 2 spaces. This is a big PR :)

File: gradebookng/tool/src/webapp/scripts/gradebook-gbgrade-table.js [hunk 16]
```diff
@@ -3277,16 +3248,18 @@ GbGradeTable.syncScore = function(studentId, assignmentId, value) {
...
+  const tableRow = GbGradeTable.rowForStudent(studentId);
 
-    // update model
-    var modelRow = GbGradeTable.modelIndexForStudent(studentId);
-    GbGradeTable.grades[modelRow][GbGradeTable.COURSE_GRADE_COLUMN_INDEX] = courseGradeData;
+  GbGradeTable.setCellState('synced', tableRow, GbGradeTable.COURSE_GRADE_COLUMN_INDEX);
+
+  const allRows = GbGradeTable.instance.getRows();
+
+  const row = allRows[tableRow];
```
Comment (adrianfish) [inline gradebookng/tool/src/webapp/scripts/gradebook-gbgrade-table.js:3258]:
```suggestion
  const row = GbGradeTable.instance.getRows()[tableRow]; 
```

File: gradebookng/tool/src/webapp/scripts/gradebook-gbgrade-table.js [hunk 17]
```diff
@@ -3277,16 +3248,18 @@ GbGradeTable.syncScore = function(studentId, assignmentId, value) {
...
-    var modelRow = GbGradeTable.modelIndexForStudent(studentId);
-    GbGradeTable.grades[modelRow][GbGradeTable.COURSE_GRADE_COLUMN_INDEX] = courseGradeData;
+  GbGradeTable.setCellState('synced', tableRow, GbGradeTable.COURSE_GRADE_COLUMN_INDEX);
+
+  const allRows = GbGradeTable.instance.getRows();
+
+  const row = allRows[tableRow]; 
+  row.update({ [GbGradeTable.COURSE_GRADE_COLUMN_INDEX]: courseGradeData });
+
+  var modelRow = GbGradeTable.modelIndexForStudent(studentId);
```
Comment (adrianfish) [inline gradebookng/tool/src/webapp/scripts/gradebook-gbgrade-table.js:3261]:
```suggestion
  const modelRow = GbGradeTable.modelIndexForStudent(studentId);
```

File: gradebookng/tool/src/webapp/scripts/gradebook-gbgrade-table.js [hunk 18]
```diff
@@ -3298,7 +3271,6 @@ GbGradeTable.syncCategoryAverage = function(studentId, categoryId, categoryScore
     var tableCol = GbGradeTable.colForCategoryScore(categoryId);
     if (tableCol >= 0) { // column is visible?
```
Comment (adrianfish) [inline gradebookng/tool/src/webapp/scripts/gradebook-gbgrade-table.js:3272]:
```suggestion
    if (GbGradeTable.colForCategoryScore(categoryId) >= 0) { // column is visible?
```

File: gradebookng/tool/src/webapp/scripts/gradebook-gbgrade-table.js [hunk 19]
```diff
@@ -3348,58 +3316,52 @@ GbGradeTable.setScore = function(studentId, assignmentId, oldScore, newScore) {
       oldScore: oldScore,
       newScore: newScore
     };
-
     var row = GbGradeTable.rowForStudent(studentId);
     var assignment = GbGradeTable.colModelForAssignment(assignmentId);
     var col = GbGradeTable.colForAssignment(assignmentId);
```
Comment (adrianfish) [inline gradebookng/tool/src/webapp/scripts/gradebook-gbgrade-table.js:3321]:
```suggestion
    const row = GbGradeTable.rowForStudent(studentId);
    const assignment = GbGradeTable.colModelForAssignment(assignmentId);
    const col = GbGradeTable.colForAssignment(assignmentId);
```

File: gradebookng/tool/src/webapp/scripts/gradebook-gbgrade-table.js [hunk 20]
```diff
@@ -3462,101 +3424,115 @@ GbGradeTable.positionModalAtTop = function($modal) {
...
+  };
 
-GbGradeTable.setupSectionsColumn = function () {
+  GbGradeTable._fixedColumns.splice(1, 0, {
+    titleFormatter: GbGradeTable.headerFormatter(GbGradeTable.STUDENTS_NUMBER_COLUMN_INDEX, 'studentNumberHeader'),
+    formatter: GbGradeTable.studentNumberCellFormatter,
+    formatterParams: {
+      _data_: GbGradeTable.students.map(function(student) {
+        return student.studentNumber || "";
+    }),
```
Comment (adrianfish) [inline gradebookng/tool/src/webapp/scripts/gradebook-gbgrade-table.js:3475]:
```suggestion
      }),
```

File: gradebookng/tool/src/webapp/scripts/gradebook-gbgrade-table.js [hunk 21]
```diff
@@ -3462,101 +3424,115 @@ GbGradeTable.positionModalAtTop = function($modal) {
...
+    columnType: "studentnumber"
```
Comment (adrianfish) [inline gradebookng/tool/src/webapp/scripts/gradebook-gbgrade-table.js:3476]:
```suggestion
      columnType: "studentnumber",
```

File: gradebookng/tool/src/webapp/scripts/gradebook-gbgrade-table.js [hunk 22]
```diff
@@ -3462,101 +3424,115 @@ GbGradeTable.positionModalAtTop = function($modal) {
...
+    const rowData = row.getData();
+    const cellKey = GbGradeTable.cleanKey(`${rowIndex}_${colIndex}_${rowData.studentId}`);
+    const wasInitialized = td.dataset.cellInitialized;
 
-        $td.attr("scope", "row").attr("role", "rowHeader");
-
-        var cellKey = (row + '_' + col);
+    if (wasInitialized === cellKey) {
+        return;
+    }
```
Comment (adrianfish) [inline gradebookng/tool/src/webapp/scripts/gradebook-gbgrade-table.js:3500]:
```suggestion
    if (wasInitialized === cellKey) return;
```

File: gradebookng/tool/src/webapp/scripts/gradebook-gbgrade-table.js [hunk 23]
```diff
@@ -3462,101 +3424,115 @@ GbGradeTable.positionModalAtTop = function($modal) {
...
-            sections: value
-        };
+    const data = Object.assign({}, { settings: GbGradeTable.settings }, { sections: value });
 
-        var html = GbGradeTable.templates.sectionsCell.setHTML(td, data);
+    if (!wasInitialized) {
+        GbGradeTable.templates.sectionsCell.setHTML(td, data);
+    } else if (wasInitialized !== cellKey) {
+        const valueCell = td.querySelector('.gb-value');
+        GbGradeTable.replaceContents(valueCell, document.createTextNode(value || ''));
```
Comment (adrianfish) [inline gradebookng/tool/src/webapp/scripts/gradebook-gbgrade-table.js:3508]:
```suggestion
        GbGradeTable.replaceContents(td.querySelector('.gb-value'), document.createTextNode(value || ''));
```

**Review by adrianfish — CHANGES_REQUESTED** [review]
Just a few small changes :)

**Review by adrianfish — DISMISSED** [review]
Just a few small changes :)

File: gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/component/GbGradeTable.html [hunk 24]
```diff
@@ -303,12 +303,12 @@
         </script>
         <script id="cellTemplate" type="text/template">
             <div class="relative">
-                <span class="gb-value">${value}</span>
+                    <span class="gb-value d-flex mt-3 justify-content-center">${value}</span>
```
Comment (jesusmmp) [inline gradebookng/tool/src/java/org/sakaiproject/gradebookng/tool/component/GbGradeTable.html:306]: Spacing

**Comment by jesusmmp** [comment]
@kunaljaykam col-index should be not be shown on the frontend

image

**Comment by jesusmmp** [comment]
It's not clear the line-through...

image

**Comment by jesusmmp** [comment]
When clicking on some icon

image


image

**Comment by jesusmmp** [comment]
When clicking on lock

image

**Comment by jesusmmp** [comment]
When clicking on x

image

**Comment by jesusmmp** [comment]
This should be float: right;

image

**Comment by jesusmmp** [comment]
Students counter is wrong

image

**Comment by jesusmmp** [comment]
Wrong styles when clicking gear icon

image

**Comment by jesusmmp** [comment]
Now, you have to do double click on each cell to insert a grade... it's more tedious

**Comment by jesusmmp** [comment]
When you enter a comment

image

**Comment by jesusmmp** [comment]
Icons gear and comments are overlap

image

**Comment by jesusmmp** [comment]
Hide category column

image

**Comment by jesusmmp** [comment]
When clicking on "Filter students" search

image

**Comment by kunaljaykam** [comment]
Thank you, @adrianfish and @jesusmmp, for your review. I was on a break but am back today. I'll be addressing and responding to all the issues you've mentioned in this PR.

**Comment by kunaljaykam** [comment]
> Now, you have to do double click on each cell to insert a grade... it's more tedious

Thank you for testing! The double-click is required due to Tabulator’s limitation when embedding both a button and an editor in the same cell. In Handsontable, it's possible to achieve single-click functionality for both. I've tried other methods to work around this limitation in Tabulator, I spent days on this but they introduced performance issues.

**Comment by kunaljaykam** [comment]
> Icons gear and comments are overlap
> 
> image

@jesusmmp Did you re-deploy the library as well?

**Comment by jesusmmp** [comment]
> > Icons gear and comments are overlap
> > image
> 
> @jesusmmp Did you re-deploy the library as well?

I'll do another test round next week!.

Great work! :)

**Comment by jesusmmp** [comment]
> > Now, you have to do double click on each cell to insert a grade... it's more tedious
> 
> Thank you for testing @jesusmmp ! The double-click is required due to Tabulator’s limitation when embedding both a button and an editor in the same cell. In Handsontable, it's possible to achieve single-click functionality for both. I've tried other methods to work around this limitation in Tabulator, I spent days on this but they introduced performance issues.

Oh!. This is really a big problem!

**Comment by ottenhoff** [comment]
Super impressed with the work so far on this. It looks good and this is super encouraging. Big thanks to all testers especially @jesusmmp !

**Comment by kunaljaykam** [comment]
> > > Now, you have to do double click on each cell to insert a grade... it's more tedious
> > 
> > 
> > Thank you for testing @jesusmmp ! The double-click is required due to Tabulator’s limitation when embedding both a button and an editor in the same cell. In Handsontable, it's possible to achieve single-click functionality for both. I've tried other methods to work around this limitation in Tabulator, I spent days on this but they introduced performance issues.
> 
> Oh!. This is really a big problem!

I have fixed it. Please test it if you get time.

**Comment by ottenhoff** [comment]
The single click is working great for me. Nice job!

In terms of performance with Chrome on my localhost with MariaDB, 41 columns * 180 students = 7380 gb scores

* Handsontable: 793ms scripting and 281ms Rendering
* Tabulator: 609ms scripting and 202ms Rendering

**Review by ern — APPROVED** [review]

**Comment by jesusmmp** [comment]
I'll do another round of testing! Thanks @kunaljaykam @ern @ottenhoff

**Comment by ottenhoff** [comment]
Thanks all for the awesome testing and the many hours of work that everyone put into this. I'm impressed! Big thanks @kunaljaykam

**Pull request merged**